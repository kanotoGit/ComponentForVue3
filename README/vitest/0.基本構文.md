## 基本構文

```ts
import { describe, it, expect } from 'vitest'

import { mount } from '@vue/test-utils'
import TextField from './TextField.vue'

describe('コンポーネント名', () => {
  it('テスト名／概要', () => {
    const wrapper = mount(TextField, { props: { msg: 'Hello Vitest' } })
    expect(wrapper.text()).toContain('Hello Vitest')
  }, 1000)
})
```