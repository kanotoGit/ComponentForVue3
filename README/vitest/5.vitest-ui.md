# vitest ui

https://vitest.dev/guide/ui.html

## インストール

```bash
yarn add @vitest/ui
```

## 実行

```bash
yarn vitest --ui
```