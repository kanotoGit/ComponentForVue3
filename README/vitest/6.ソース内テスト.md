# ソース内テスト

https://vitest.dev/guide/in-source.html

小さいプロジェクトなどで有用
まぁ、使わない