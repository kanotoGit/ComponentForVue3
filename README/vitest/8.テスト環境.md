# テスト環境

`jsdom`使うか`happy-dom`使うかみたいな感じで使用するライブラリ（環境）を変更できる
こだわりなければ、たぶん使わない？