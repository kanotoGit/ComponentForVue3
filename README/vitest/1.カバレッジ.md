# カバレッジ

各テストがどれだけ実行されたかを閲覧できる機能.
https://vitest.dev/guide/coverage.html

## インストール

```bash
yarn add @vitest/coverage-c8
```

```ts
// vite.config.ts

import { defineConfig } from 'vitest/config'

export default defineConfig({
  test: {
    coverage: {
      provider: 'c8' // or 'istanbul'
    },
  },
})
```


## 実行

```json
// package.json

{
  "scripts": {
    "test:coverage": "vitest run --coverage",
  }
}
```

```bash
yarn test:coverage
```

## 実行結果に出力方法

```ts
// vitest.config.ts

export default mergeConfig(
  defineConfig({
    test: {
      // カバレッジ
      coverage: {
        provider: 'c8', // or 'istanbul'
        reporter: ['text', 'html', 'clover', 'json'],
      },
    }
  })
)
```

## カスタムカバレッジプロバイダー
## デフォルトのカバレッジフォルダーの場所の変更
## コードを無視する

## その他機能

https://vitest.dev/config/#coverage