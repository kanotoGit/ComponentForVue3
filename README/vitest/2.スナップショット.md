# スナップショット

UI が予期せず変更されていないかを確かめるのに非常に有用なツールです。
https://jestjs.io/ja/docs/snapshot-testing
https://developer.mamezou-tech.com/testing/jest/jest-snapshot-testing/

## 記述方法

```ts
import { expect, it } from 'vitest'

it('toUpperCase', () => {
  const result = 'foobar'.toLocaleUpperCase()
  expect(result).toMatchSnapshot()
})
```


## 実行

vitestを実行すると[__snapshots__]フォルダが生成される
その中に実行結果（スナップショットのコード）が生成される
※生成ごとにコミット＆プッシュしてレビュー対象にする


## 更新

既に生成されているスナップショットを更新したい場合に使う

```bash
vitest -u
```


## インラインスナップショット

テスト結果をインライン上（specファイル上）で記述できる
※あまり使わないかも


## ファイルのスナップショット

ファイルの内容とテスト結果を比較できるやつ
※あまり使わないかも

## 画像のスナップショット

画像の内容とテスト結果を比較できるやつ
※あまり使わないかも

## カスタムシリアライザー