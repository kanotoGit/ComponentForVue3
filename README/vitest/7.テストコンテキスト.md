# テストコンテキスト

パラメータ追加などを行う
※用途詳細不明
https://vitest.dev/guide/test-context.html


## コンテキストの定義(グローバル)

```ts
// ./src/vitest.d.ts
import vitest from 'vitest'

declare module 'vitest' {
  interface TestContext {
    foo?: string
  }
}
```

```ts
// ./src/xxxx.ts

/** テスト開始直前に実行 */
beforeEach<LocalTestContext>(async (context) => {
  // typeof context is 'TestContext & LocalTestContext'
  context.foo = 'bar'
})
```


## コンテキストの定義(ローカル)

```ts
/** コンテキストにパラメータの追加 */
interface LocalTestContext {
  foo: string
}
beforeEach<LocalTestContext>(async (context) => {
  // typeof context is 'TestContext & LocalTestContext'
  context.foo = 'bar'
})

it<LocalTestContext>('should work', ({ foo }) => {
  // typeof foo is 'string'
  console.log(foo) // 'bar'
})
```