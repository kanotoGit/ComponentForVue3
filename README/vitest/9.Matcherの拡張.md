# matcherの拡張

独自のテスト判定（`toBeXXXXX`）を実装できる


## 判定の定義

### 判定ファイル
```ts
// ./src/vitest.setup.ts

import { expect } from "vitest"

expect.extend({
  toBeFoo(received, expected) {
    console.log(received, expected);

    const { isNot } = this
    return {
      pass: received === 'foo',
      message: () => `${received} is${isNot ? ' not' : ''} foo`
    }
  }
})
```

### 判定ファイルの読み込み
```ts
// ./vitest.config.ts
export default mergeConfig(
  defineConfig({
    test: {
      // セットアップ
      setupFiles: [
        './src/vitest.setup.ts'
      ]
    }
  })
)
```

### 型
```ts
// ./src/vitest.d.ts
import vitest from 'vitest'

interface CustomMatchers<R = unknown> {
  // 独自のmatcher
  toBeFoo(): R
}

declare module 'vitest' {
  interface Assertion<T = any> extends CustomMatchers<T> {}
  interface AsymmetricMatchersContaining extends CustomMatchers {}
}
```


## 判定実行

```ts
// XXXX.test.ts
it('xxxxxx', async (context) => {
  expect('foo').toBeFoo()
})
```