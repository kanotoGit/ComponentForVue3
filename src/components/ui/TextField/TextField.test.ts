import { describe, it, expect, type TestOptions, expectTypeOf, assertType, type TestContext, beforeEach } from 'vitest'

import { mount } from '@vue/test-utils'
import TextField from './TextField.vue'

/** コンテキストにパラメータの追加 */
interface LocalTestContext {
  foo: string
}
beforeEach<LocalTestContext>(async (context) => {
  // typeof context is 'TestContext & LocalTestContext'
  context.foo = 'bar'
})

describe('TextField', () => {
  // it('renders properly', () => {
  //   const wrapper = mount(TextField, { props: { text: 'Hello Vitest' } })
  //   expect(wrapper.text()).to('Hello Vitest')
  // }, { timeout: 1000, repeats: 5, retry: 1 })

  // it('toUpperCase', () => {
  //   const result = 'foobar2'.toLocaleUpperCase()
  //   expect(result).toMatchSnapshot()
  //   expectTypeOf(result).toEqualTypeOf<string>()
  // })

  // it<LocalTestContext>('コンテキストのサンプル', (context: LocalTestContext) => {
  //   console.log(context.foo)
  // })

  // it('matcher拡張のサンプル', () => {
  //   // matcher拡張
  //   expect('foo').toBeFoo()
  // })

  it('入力値が数値の場合、number型がemitされること', async () => {
    const wrapper = mount(TextField, {
      props: {
        text: null,
        type: 'number',
      },
    });

    // 値の入力
    const input = wrapper.find('input');
    await input.setValue('123');

    // update:textのチェック
    const updateTextEmit = wrapper.emitted('update:text')
    expect(updateTextEmit).toBeTruthy();
    expect(updateTextEmit?.[0]?.[0]).toBe(123)

    // inputイベントのチェック
    const inputEmit = wrapper.emitted('input')
    expect(inputEmit).toBeTruthy();
    expect(inputEmit?.[0]?.[0]).toBe(123)
    expect(inputEmit?.[0]?.[1]).toBeInstanceOf(Event)

    // changeイベントのチェック
    const changeEmit = wrapper.emitted('change')
    expect(changeEmit).toBeTruthy();
    expect(changeEmit?.[0]?.[0]).toBe(123)
    expect(changeEmit?.[0]?.[1]).toBeInstanceOf(Event)

    expect(input).toMatchSnapshot()
  })
})
