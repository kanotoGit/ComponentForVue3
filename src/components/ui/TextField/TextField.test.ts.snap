// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextField > toUpperCase 1`] = `"FOOBAR2"`;

exports[`TextField > 入力値が数値の場合、number型がemitされること 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <input
    data-v-66836ec0=""
    type="text"
  />,
}
`;
